File: parser_bot.py
#!/usr/bin/env python3
import json
import re
import sys
from typing import Optional, Dict

INTENT_RULES = {
    'ajuda': [r'\b(ajuda|help|/start|oi|ol[aá]|ola|bom dia|boa tarde|boa noite)\b'],
    'devolucao_fornecedor': [r'\b(devolu[cç][aã]o|devolvendo|devolvendo pro|devolvendo para|devolucao)\b', r'\b(fornecedor)\b'],
    'voltou_estoque': [r'\b(voltou|retornou|retorno|dispon[íi]vel|disponivel|de volta|de volta ao estoque|retornou ao estoque)\b'],
    'usado_no_carro': [r'\b(usei|usado|retirado|retirada|aplicado|instalado|coloquei|utilizei)\b'],
    'buscar_peca': [r'\b(onde est[áa]|onde esta|cad[eê]|procura|procurei|localiza|localizar|saber onde|onde fica)\b']
}

VEHICLE_WORDS = {
    'cobalt', 'renegade', 'tcross', 'hb20', 'onix', 'fiesta', 'focus', 'gol', 'palio', 'uno',
    'ka', 'creta', 'city', 'civic', 'corolla', 'prisma', 'fox', 'jeep', 's10', 'hilux'
}

KNOWN_PARTS = [
    'lanterna', 'sensor', 'moldura', 'parachoque', 'farol', 'bateria', 'amortecedor', 'pastilha',
    'disco', 'embreagem', 'filtro', 'bomba', 'radiador', 'pneu', 'suspensao', 'escape', 'capo',
    'painel', 'porta', 'vidro', 'mola', 'corrente', 'engrenagem'
]

IGNORE_WORDS = {
    'onde', 'esta', 'esta', 'o', 'a', 'um', 'uma', 'de', 'do', 'da', 'no', 'na', 'por', 'pro',
    'para', 'com', 'sem', 'foi', 'me', 'minha', 'meu', 'peca', 'peça', 'veiculo', 'carro', 'peças'
}

PLATE_TOKEN = re.compile(r'\b([A-Za-z0-9-]{4,8})\b')


def normalize(text: str) -> str:
    text = text or ''
    text = text.strip()
    text = text.lower()
    # Convert accented characters
    text = re.sub(r'[áàâãä]', 'a', text)
    text = re.sub(r'[éèêë]', 'e', text)
    text = re.sub(r'[íìîï]', 'i', text)
    text = re.sub(r'[óòôõö]', 'o', text)
    text = re.sub(r'[úùûü]', 'u', text)
    text = re.sub(r'[ç]', 'c', text)
    return text


def extract_plate(text: str) -> Optional[str]:
    text = text or ''
    for token in PLATE_TOKEN.findall(text):
        candidate = re.sub(r'[^A-Za-z0-9]', '', token).upper()
        if len(candidate) < 4 or len(candidate) > 8:
            continue
        # Valid pattern 3 letters + 4 digits or 4 letters + 3 digits or mixed alnum
        if re.fullmatch(r'[A-Z]{3}\d{4}|[A-Z]{4}\d{3}|[A-Z0-9]{4,8}', candidate):
            return candidate
    return None


def extract_os(text: str) -> Optional[str]:
    text = text or ''
    patterns = [
        r'\bOS\s*[:\-]?\s*(\d{2,})\b',
        r'\bos\s*(?:numero|n[ºo]|#)?\s*[:\-]?\s*(\d{2,})\b',
        r'\bordem de servico\s*(?:numero)?\s*[:\-]?\s*(\d{2,})\b'
    ]
    for p in patterns:
        m = re.search(p, text, flags=re.IGNORECASE)
        if m:
            return m.group(1)
    return None


def classify_intent(raw_text: str) -> str:
    normalized = normalize(raw_text)
    if not normalized:
        return 'desconhecido'

    for intent, regexes in INTENT_RULES.items():
        if intent == 'devolucao_fornecedor':
            # require both general word and fornecedor mention
            if all(re.search(r, normalized) for r in regexes):
                return 'devolucao_fornecedor'
            continue
        if any(re.search(r, normalized) for r in regexes):
            return intent
    return 'desconhecido'


def parse_piece(raw_text: str) -> Optional[str]:
    if not raw_text:
        return None
    lower = normalize(raw_text)
    for piece in KNOWN_PARTS:
        if re.search(rf'\b{re.escape(piece)}\b', lower):
            return piece.capitalize()

    tokens = re.findall(r"[a-z0-9]+", lower)
    for token in tokens:
        if token in IGNORE_WORDS or token in VEHICLE_WORDS:
            continue
        if token.isdigit():
            continue
        if len(token) <= 2:
            continue
        return token.capitalize()
    return None


def extract_reason(raw_text: str) -> Optional[str]:
    if not raw_text:
        return None
    lower = raw_text.strip()
    lower_norm = normalize(raw_text)
    if 'veio errado' in lower_norm:
        return 'veio errado'
    m = re.search(r'(?:motivo\s*[:\-]?\s*|porque\s+|pq\s+|pois\s+)(.+)', lower, flags=re.IGNORECASE)
    if m:
        reason = m.group(1).strip()
        if reason:
            return reason
    return None


def parse_message(raw_text: str) -> Dict[str, Optional[str]]:
    if not raw_text or not raw_text.strip():
        return {
            'intent': 'desconhecido',
            'peca': None,
            'peca_secundaria': None,
            'placa': None,
            'responsavel': None,
            'os_numero': None,
            'motivo': None
        }

    intent = classify_intent(raw_text)
    placa = extract_plate(raw_text)
    os_numero = extract_os(raw_text)
    peca = parse_piece(raw_text)
    motivo = extract_reason(raw_text)

    return {
        'intent': intent,
        'peca': peca,
        'peca_secundaria': None,
        'placa': placa,
        'responsavel': None,
        'os_numero': os_numero,
        'motivo': motivo
    }


def main() -> None:
    arg_text = ' '.join(sys.argv[1:]).strip()
    if not arg_text:
        print('Uso: python parser_bot.py "mensagem do cliente"')
        print('Exemplo: python parser_bot.py "onde está a lanterna renegade GGL8443"')
        sys.exit(0)

    parsed = parse_message(arg_text)
    print(json.dumps(parsed, ensure_ascii=False, indent=2))


if __name__ == '__main__':
    main()

File: parserBot.js
const intentKeywords = {
  ajuda: [/\b(ajuda|help|\/start|oi|olá|ola|bom dia|boa tarde|boa noite)\b/i],
  devolucao_fornecedor: [/\b(devolu[cç][aã]o|devolvendo|devolvendo pro|devolvendo para|devolucao)\b/i, /\b(fornecedor|fornecedor)\b/i],
  voltou_estoque: [/\b(voltou|retornou|retorno|dispon[íi]vel|disponivel|de volta|devolvido ao estoque|devolvida ao estoque)\b/i],
  usado_no_carro: [/\b(usei|usado|retirado|retirada|aplicado|instalado|coloquei|utilizei)\b/i],
  buscar_peca: [/\b(onde est[áa]|onde esta|cad[eê]|procura|procurei|localiza|saber onde)\b/i]
};

const vehicleWords = [
  'cobalt','renegade','tcross','hb20','onix','fiesta','focus','gol','palio','uno','ka','creta','city','civic','corolla'
];

function normalizeText(text) {
  return text.normalize('NFKD').replace(/[\u0300-\u036f]/g, '').toLowerCase();
}

function extractPlate(raw) {
  if (!raw) return null;
  // Remove dashes and spaces and convert to uppercase
  const cleaned = raw.toUpperCase().replace(/[-\s]/g, '');
  // Brazilian plate heuristics: 3 letters + 4 digits OR 4 letters + 3 digits
  if (/^[A-Z]{3}\d{4}$/.test(cleaned) || /^[A-Z]{4}\d{3}$/.test(cleaned) || /^[A-Z0-9]{4,8}$/.test(cleaned)) {
    return cleaned;
  }
  return null;
}

function findPlate(text) {
  if (!text) return null;
  // Find tokens looking like plates
  const tokens = text.match(/\b[\w-]{4,8}\b/g);
  if (!tokens) return null;
  for (const token of tokens) {
    const candidate = token.replace(/[^A-Za-z0-9]/g, '');
    const plate = extractPlate(candidate);
    if (plate) return plate;
  }
  return null;
}

function findOS(text) {
  if (!text) return null;
  const m = text.match(/\bOS\s*[:\-]?\s*(\d{2,})\b/i);
  if (m) return m[1];
  const m2 = text.match(/\b(?:OS|os|os\s+numero|nº\s*OS)\s*(\d{2,})\b/i);
  if (m2) return m2[1];
  return null;
}

function parsePiece(text) {
  if (!text) return null;
  // common part words and stop words
  const knownParts = [
    'lanterna','sensor','moldura','parachoque','farol','bateria','amortecedor','pastilha','disco','embreagem','filtro','bomba','radiador','pneu','suspensao','escape','capo','painel','porta','vidro'
  ];
  const cleaned = normalizeText(text);
  for (const part of knownParts) {
    const regex = new RegExp('\\b'+part+'\\b', 'i');
    if (regex.test(cleaned)) {
      return part.charAt(0).toUpperCase() + part.slice(1);
    }
  }
  // fallback: pick first non-stop word around keywords
  const ignore = new Set(['onde','esta','esta','o','a','um','uma','de','do','da','no','na','por','pro','para','com','sem','foi','foi','usado','usei','voltou','voltado','retornou','peca','peça','peça']);
  const tokens = cleaned.split(/[^a-z0-9]+/).filter(Boolean);
  for (const t of tokens) {
    if (ignore.has(t)) continue;
    if (vehicleWords.includes(t)) continue;
    if (/^\d+$/.test(t)) continue;
    if (t.length <= 2) continue;
    return t.charAt(0).toUpperCase() + t.slice(1);
  }
  return null;
}

function classifyIntent(rawText) {
  if (!rawText || !rawText.trim()) return 'desconhecido';
  const text = normalizeText(rawText);
  if (intentKeywords.ajuda.some(rx => rx.test(text))) return 'ajuda';
  if (intentKeywords.devolucao_fornecedor.every(rx => rx.test(text))) return 'devolucao_fornecedor';
  if (intentKeywords.voltou_estoque.some(rx => rx.test(text))) return 'voltou_estoque';
  if (intentKeywords.usado_no_carro.some(rx => rx.test(text))) return 'usado_no_carro';
  if (intentKeywords.buscar_peca.some(rx => rx.test(text))) return 'buscar_peca';
  return 'desconhecido';
}

function parseMessage(rawText) {
  if (!rawText || !rawText.trim()) {
    return {
      intent: 'desconhecido',
      peca: null,
      peca_secundaria: null,
      placa: null,
      responsavel: null,
      os_numero: null,
      motivo: null
    };
  }

  const intent = classifyIntent(rawText);
  const placa = findPlate(rawText);
  const os_numero = findOS(rawText);

  // tentativa simples de extrair motivo se for devolucao fornecedor
  let motivo = null;
  const m = rawText.match(/(?:motivo de|motivo:|porque|pq|pois|veio)\s+(.+)/i);
  if (m) {
    motivo = m[1].trim();
  } else if (/\bveio errado\b/i.test(rawText)) {
    motivo = 'veio errado';
  }

  const peca = parsePiece(rawText);

  return {
    intent,
    peca: peca || null,
    peca_secundaria: null,
    placa: placa || null,
    responsavel: null,
    os_numero: os_numero || null,
    motivo: motivo || null
  };
}

if (require.main === module) {
  const input = process.argv.slice(2).join(' ');
  if (!input) {
    console.log('Uso: node parserBot.js "texto da mensagem"');
    console.log('Exemplo: node parserBot.js "onde está a lanterna renegade GGL8443"');
    process.exit(0);
  }
  console.log(JSON.stringify(parseMessage(input), null, 2));
}

module.exports = { parseMessage };
